### PR TITLE
Add DaemonSet to enable kdump for ubuntu

### PR DESCRIPTION
**Pull request by bobbypage**

### CODE REVIEW

File: enable-kdump/ubuntu-enable-kdump.yaml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                  echo "${kdump_show}"
+                  return 0
+              }
+
+              function install() {
+                  echo "installing kdump"
+                  apt-get update
+                  DEBIAN_FRONTEND=noninteractive apt-get install -y kdump-tools
+                  echo "kdump enabled; rebooting..."
+                  sleep 30
```
Comment (karan) [inline enable-kdump/ubuntu-enable-kdump.yaml]: why sleep? maybe turn this down to something smaller like 10?

File: enable-kdump/ubuntu-enable-kdump.yaml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                  else
+                      echo "kdump not setup, isn't ready"
+                  fi
+                  echo "${kdump_show}"
+              }
+
+              function install() {
+                  echo "installing kdump"
+                  apt-get update
+                  DEBIAN_FRONTEND=noninteractive apt-get install -y kdump-tools
```
Comment (karan) [inline enable-kdump/ubuntu-enable-kdump.yaml]: I tried the defaults once mistakenly for kdump-tools and the reboot was borked. I believe the first question answered yes causes an interactive reboot (with username/password). when i answered no to the first question, then it does work. did you test this?

File: enable-kdump/ubuntu-enable-kdump.yaml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                  echo "kdump enabled; rebooting..."
```
Comment (karan) [inline enable-kdump/ubuntu-enable-kdump.yaml]: You also want to enable hung task panic

```
echo 1 > /proc/sys/kernel/hung_task_panic"
echo 20 > /proc/sys/kernel/hung_task_timeout_secs"
```

EDIT: actually, maybe also set the timeout here so it persists at reboot

Comment (bsdnet) [reply enable-kdump/ubuntu-enable-kdump.yaml]: You need to change /etc/sysctl files if you want hung triggered kdump automatically.

File: enable-kdump/ubuntu-enable-kdump.yaml (hunk 1 above)
Comment (bowei) [reply enable-kdump/ubuntu-enable-kdump.yaml]: yeah -- what are we waiting for? if it's an active thing, can we have it just inf loop printing the status.

like

Waiting for reboot (x seconds)...
Waiting for reboot (x seconds)...
Waiting for reboot (x seconds)...
Waiting for reboot (x seconds)...
Waiting for reboot (x seconds)...

(NO CARRIER)

File: enable-kdump/ubuntu-enable-kdump.yaml (hunk 3 above)
Comment (karan) [reply enable-kdump/ubuntu-enable-kdump.yaml]: Oh yeah:

```
echo 1 > /proc/sys/kernel/sysrq
```

File: enable-kdump/ubuntu-enable-kdump.yaml (hunk 2 above)
Comment (bowei) [reply enable-kdump/ubuntu-enable-kdump.yaml]: stack overflow is particularly unhelpful on this question...

Comment (karan) [reply enable-kdump/ubuntu-enable-kdump.yaml]: [This](https://ubuntu.com/server/docs/kernel-crash-dump) might work (modify files after defaults):

```
# File /etc/default/kexec
# Load a kexec kernel (true/false)
LOAD_KEXEC=false

# File /etc/default/kdump-tools
USE_KDUMP=1
```

Comment (bobbypage) [reply enable-kdump/ubuntu-enable-kdump.yaml]: The docs for setting up kdump for ubuntu are here: https://ubuntu.com/server/docs/kernel-crash-dump

The two prompts it shows are first if you want to hook up kexec-tools for future reboots and if kdump-tools should be enabled by default. Ubuntu docs say yes should be answer to both questions. 

From my testing and stackoverflow searching, the `DEBIAN_FRONTEND=noninteractive` makes it so the prompts aren't shown and the -y should answer yes to all the prompts.

I tested it and on the next reboot, `kdump-config show`, showed "ready to dump", and `/proc/cmdline` had the reserved memory for crashkernel

```
$ cat /proc/cmdline
BOOT_IMAGE=/boot/vmlinuz-5.0.0-1032-gke root=PARTUUID=f56597fd-2626-41aa-a805-9572e5aced4a ro scsi_mod.use_blk_mq=Y console=ttyS0 net.ifnames=0 crashkernel=512M-:192M
```

based on that I assumed it's setup correctly :). We can trigger a panic and ensure the dump is generated to double check as a test.

File: enable-kdump/ubuntu-enable-kdump.yaml (hunk 3 above)
Comment (bobbypage) [reply enable-kdump/ubuntu-enable-kdump.yaml]: Thanks good point. Setup the sysctls for `hung_task_panic` and `hung_task_timeout_secs`, and made it so they are applied on next reboot after kdump is installed.

@bsdnet by change `/etc/sysctl` you mean apply the above sysctls or are you referring to some other sysctls?

@karan why do we need to enable sysrq? I don't think it's required for kdump, unless we want to trigger kdump as a test by forcing a panic via sysrq.

File: enable-kdump/ubuntu-enable-kdump.yaml (hunk 1 above)
Comment (bobbypage) [reply enable-kdump/ubuntu-enable-kdump.yaml]: The sleep is just so we don't do the reboot right away and can ensure we have enough time for the logs from the script to be flushed, etc. Changed to 10 secs, which should be fine. 

Added "waiting for reboot", sadly the "NO CARRIER" will have to be client side :)

File: enable-kdump/ubuntu-enable-kdump.yaml (hunk 3 above)
Comment (karan) [reply enable-kdump/ubuntu-enable-kdump.yaml]: Oh I misread the sysctl comment as sysrq. Disregard.

File: enable-kdump/ubuntu-enable-kdump.yaml (hunk 2 above)
Comment (karan) [reply enable-kdump/ubuntu-enable-kdump.yaml]: From my understanding kdump uses kexec to boot a capture kernel. So by making kexec and kdump both handle reboots, are we causing a potential conflict?

Can you try to take a capture on a GKE Node following your script and see if you can load the vmcore using `crash`?

Comment (bobbypage) [reply enable-kdump/ubuntu-enable-kdump.yaml]: yeah, as I understand from [1] and [2] the way kdump works is to loads the capture kernel ("dump capture kernel"). When a crash happens, as I understand the kdump uses kexec to boot up the capture kernel which captures the dump. As I understand, I think both need to enabled here... let me try full test with the crash tool.   

[1] https://wiki.archlinux.org/index.php/Kdump
[2] https://www.kernel.org/doc/html/latest/admin-guide/kdump/kdump.html

**Review by bowei — APPROVED** [review]
/lgtm

File: enable-kdump/ubuntu-enable-kdump.yaml (hunk 2 above)
Comment (karan) [reply enable-kdump/ubuntu-enable-kdump.yaml]: Were you able to get it tested?

Comment (bobbypage) [reply enable-kdump/ubuntu-enable-kdump.yaml]: Yup, I tested it by triggering crash with sysrq and used `crash` tool to look at the resulting dump. I added a README file to document the process a bit as well.

**Pull request merged**